### PR TITLE
Add initial version for the new testing framework

### PR DESCRIPTION
**Pull request by Neverlord**
@riemass this a cleaned up version of the prototype we've looked at recently. A few of the details have changed, but it's overall still the same beast. Of course, I don't expect you do go over all of the new lines. There's no design document, but I hope everything makes sense with the discussion we had recently. We can also schedule another session if some things don't add up for you.

This is not feature complete yet, but I think it's a good enough state to pull it into the repo and start iterating on it.

### CODE REVIEW

File: libcaf_test/src/test/context.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  reporter::instance->end_step(ptr);
+}
+
+bool context::activated(block* ptr) const noexcept {
+  return std::find(path.begin(), path.end(), ptr) != path.end();
+}
+
+bool context::can_run() {
+  auto pred = [](auto& kvp) { return kvp.second->can_run(); };
+  return !steps.empty() && std::any_of(steps.begin(), steps.end(), pred);
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_test/src/test/context.cpp:33]: You're right, that redundant. I'll drop it.

**Comment by Neverlord** [comment]
> From what I can see, there is a cli option for regex suite selection, but the feature is not implemented yet?

Correct. I have added the option already because our `ctest` setup (and thus CI) otherwise breaks, but these flags still need to be implemented:

- `-n` for disabling colored output
- `-s <regex>` for selecting suites
- `-t <regex>` for selecting tests
- `-m <seconds>` for aborting the test after reaching the `<seconds>` limit

Probably the first followup on this PR.

**Comment by riemass** [comment]
Suite selection should come at least before porting a significant number of current tests. That and `require(_eq)?` would be two features I look forward to.

**Review by riemass — APPROVED** [review]
:tada:

**Pull request merged**